truffle/src/main/java/org/jruby/truffle/core/exception/ExceptionNodes.java

```diff
@@ -133,11 +133,25 @@ public AllocateNode(RubyContext context, SourceSection sourceSection) {
             allocateObjectNode = AllocateObjectNodeGen.create(context, sourceSection, null, null);
         }
 
-        @Specialization
+        @Specialization(guards = "!isNameError(rubyClass)")
         public DynamicObject allocate(DynamicObject rubyClass) {
             return allocateObjectNode.allocate(rubyClass, nil(), null, nil());
         }
 
+        @Specialization(guards = "isNameError(rubyClass)")
+        public DynamicObject allocateNameError(DynamicObject rubyClass) {
+            return allocateObjectNode.allocate(rubyClass, nil(), null, null, nil());
+        }
```

Comment:
I would define this on NameError, not in Exception so there is no need for guards and it's automatically using the right one.